---
title: "Compare VCF and text for storing GWAS summary statistics"
author: "Matt Lyon"
date: "2020-02-25"
---

### Prepare files

Download BMI GWAS summary statistics and annotation data (Neale et al)

```{r}
library('data.table')
library('stringr')

# UK Biobank â€” Neale lab. (n.d.). Retrieved February 25, 2020, from http://www.nealelab.is/uk-biobank/
download.file("https://www.dropbox.com/s/tx8iw9lk53z3vj9/21001_raw.gwas.imputed_v3.both_sexes.tsv.bgz?raw=1", destfile="/data/gwas.raw.txt.gz")
download.file("https://www.dropbox.com/s/puxks683vb0omeg/variants.tsv.bgz?raw=1", destfile="/data/lookup.txt.gz")
gwas <- fread("/data/gwas.raw.txt.gz")
lookup <- fread("/data/lookup.txt.gz")
gwas <- merge(gwas, lookup, "variant")
rm(lookup)

# save output to tab file
write.table(gwas[,c("chrom", "pos", "alt", "ref", "beta", "se", "pval", "rsid", "AF", "info", "n_complete_samples")], file="/data/gwas.txt", sep="\t", row.names=F, quote=F)

message(paste("Total number of variants in GWAS:", nrow(gwas)))
```

Define file schema for mapping plain text to VCF format

```{bash}
set -euo pipefail
gzip /data/gwas.txt
echo "{\"chr_col\": 0, \"pos_col\": 1, \"ea_col\": 2, \"oa_col\": 3, \"beta_col\": 4, \"se_col\": 5, \"pval_col\": 6, \"delimiter\": \"\t\", \"header\": true, \"snp_col\": 7, \"eaf_col\": 8, \"imp_info_col\": 9, \"ncontrol_col\": 10,  \"build\": \"GRCh37\"}" > /data/schema.json
```

Convert GWAS to VCF. This will perform on-the-fly harmonisation of the effect allele to the VCF alternative allele ensuring data consistency. Source code available here: [gwas2vcf](https://github.com/MRCIEU/gwas2vcf).

```{bash}
set -euo pipefail
python /app/gwas2vcf/main.py \
--out /data/gwas.vcf.gz \
--data /data/gwas.txt.gz \
--ref /data/human_g1k_v37.fasta \
--id "neale-bmi-2018" \
--json /data/schema.json
```

Extract files for later comparison

```{bash}
gzip -dc /data/gwas.txt.gz > /data/gwas.txt
gzip -dc /data/gwas.vcf.gz > /data/gwas.vcf
```

### Compare file sizes

Gzipped unstructured (original) text

```{bash}
set -euo pipefail
ls -lh /data/gwas.txt.gz
```

Gzipped VCF

```{bash}
set -euo pipefail
ls -lh /data/gwas.vcf.gz
```

### Compare query execution time between VCF and unstructured text

Obtain random set of SNPs for measuring query performance

```{bash}
set -eu
gzip -dc /data/gwas.txt.gz | awk 'NR>1 {print $3"\t"$1":"$2"-"$2}' | shuf | head -n 100 > /data/queries.txt
```

Record time for random lookups by rsid

```{bash}
set -euo pipefail

echo -e "query\treal\tuser\tsys" > /data/rsid.query.uncompressed.text.awk.time.txt
echo -e "query\treal\tuser\tsys" > /data/rsid.query.compressed.text.awk.time.txt
echo -e "query\treal\tuser\tsys" > /data/rsid.query.compressed.vcf.bcftools.time.txt
echo -e "query\treal\tuser\tsys" > /data/rsid.query.compressed.vcf.awk.time.txt
echo -e "query\treal\tuser\tsys" > /data/rsid.query.uncompressed.vcf.awk.time.txt

while read line; do

    # define query
    snp=$(echo "$line" | cut -s -f1)

    # log SNP
    echo -ne "$snp\t" >> /data/rsid.query.uncompressed.text.awk.time.txt
    echo -ne "$snp\t" >> /data/rsid.query.compressed.text.awk.time.txt
    echo -ne "$snp\t" >> /data/rsid.query.compressed.vcf.bcftools.time.txt
    echo -ne "$snp\t" >> /data/rsid.query.compressed.vcf.awk.time.txt
    echo -ne "$snp\t" >> /data/rsid.query.uncompressed.vcf.awk.time.txt

    # compressed text
    /usr/bin/time -f "%e\t%U\t%S" -ao /data/rsid.query.compressed.text.awk.time.txt \
    gzip -dc /data/gwas.txt.gz | \
    awk -v snp="$snp" -F"\t" '$3==snp'

    # uncompressed text
    /usr/bin/time -f "%e\t%U\t%S" -ao /data/rsid.query.uncompressed.text.awk.time.txt \
    awk -v snp="$snp" -F"\t" '$3==snp' /data/gwas.txt

    # compressed VCF using bcftools
    /usr/bin/time -f "%e\t%U\t%S" -ao /data/rsid.query.compressed.vcf.bcftools.time.txt \
    bcftools query -i "%ID == \"$snp\"" -f '%LINE\n' /data/gwas.vcf.gz

    # compressed VCF using awk
    /usr/bin/time -f "%e\t%U\t%S" -ao /data/rsid.query.compressed.vcf.awk.time.txt \
    gzip -dc /data/gwas.vcf.gz | \
    awk -v snp="$snp" -F"\t" '$3==snp'

    # uncompressed VCF using awk
    /usr/bin/time -f "%e\t%U\t%S" -ao /data/rsid.query.uncompressed.vcf.awk.time.txt \
    awk -v snp="$snp" -F"\t" '$3==snp' /data/gwas.vcf

done < /data/queries.txt
```

Plot results

```{r}
library('ggplot2')
library('data.table')

# read in data
rsid.query.uncompressed.text.awk.time <- fread("/data/rsid.query.uncompressed.text.awk.time.txt")
rsid.query.compressed.text.awk.time <- fread("/data/rsid.query.compressed.text.awk.time.txt")
rsid.query.compressed.vcf.bcftools.time <- fread("/data/rsid.query.compressed.vcf.bcftools.time.txt")
rsid.query.compressed.vcf.awk.time <- fread("/data/rsid.query.compressed.vcf.awk.time.txt")
rsid.query.uncompressed.vcf.awk.time <- fread("/data/rsid.query.uncompressed.vcf.awk.time.txt")

rsid.query.uncompressed.text.awk.time$method <- "uncompressed.text.awk"
rsid.query.compressed.text.awk.time$method <- "compressed.text.awk"
rsid.query.compressed.vcf.bcftools.time$method <- "compressed.vcf.bcftools"
rsid.query.compressed.vcf.awk.time$method <- "compressed.vcf.awk"
rsid.query.uncompressed.vcf.awk.time$method <- "uncompressed.vcf.awk"

# merge
all <- rbind(rsid.query.uncompressed.text.awk.time[, c("real", "method")], rsid.query.compressed.text.awk.time[, c("real", "method")], rsid.query.compressed.vcf.bcftools.time[, c("real", "method")], rsid.query.compressed.vcf.awk.time[, c("real", "method")], rsid.query.uncompressed.vcf.awk.time[ ,c("real", "method")])
all$method <- as.factor(all$method)

# t test -- is there a query time difference uncompressed text and uncompressed VCF using awk?
t.test(rsid.query.uncompressed.text.awk.time$real, rsid.query.uncompressed.vcf.awk.time$real, paired = TRUE, alternative = "two.sided")

# plot
ggplot(all, aes(color = method, y = real)) + geom_boxplot() + labs(x = "Method", y = "Execution time (sec)") + theme(axis.text.x=element_blank()) + theme(legend.title=element_blank()) + ggtitle("Query time using dbsnp identifier") + ylim(0, 6)
```

Record time for random lookups by chromosome and position

```{bash}
set -euo pipefail

echo -e "query\treal\tuser\tsys" > /data/chrpos.query.uncompressed.text.awk.time.txt
echo -e "query\treal\tuser\tsys" > /data/chrpos.query.compressed.text.awk.time.txt
echo -e "query\treal\tuser\tsys" > /data/chrpos.query.compressed.vcf.bcftools.time.txt
echo -e "query\treal\tuser\tsys" > /data/chrpos.query.compressed.vcf.awk.time.txt
echo -e "query\treal\tuser\tsys" > /data/chrpos.query.uncompressed.vcf.awk.time.txt

while read line; do

    # define query
    query=$(echo "$line" | cut -s -f2)
    chr=$(echo "$query" | cut -d':' -f1)
    pos=$(echo "$query" | cut -d':' -f2 | cut -d- -f1)

    # log query
    echo -ne "$query\t" >> /data/chrpos.query.uncompressed.text.awk.time.txt
    echo -ne "$query\t" >> /data/chrpos.query.compressed.text.awk.time.txt
    echo -ne "$query\t" >> /data/chrpos.query.compressed.vcf.bcftools.time.txt
    echo -ne "$query\t" >> /data/chrpos.query.compressed.vcf.awk.time.txt
    echo -ne "$query\t" >> /data/chrpos.query.uncompressed.vcf.awk.time.txt

    # compressed text
    /usr/bin/time -f "%e\t%U\t%S" -ao /data/chrpos.query.compressed.text.awk.time.txt \
    gzip -dc /data/gwas.txt.gz | \
    awk -v chr="$chr" -v pos="$pos" -F"\t" '$1 == chr && $2 == pos'

    # uncompressed text
    /usr/bin/time -f "%e\t%U\t%S" -ao /data/chrpos.query.uncompressed.text.awk.time.txt \
    awk -v chr="$chr" -v pos="$pos" -F"\t" '$1 == chr && $2 == pos' /data/gwas.txt

    # compressed VCF using bcftools
    /usr/bin/time -f "%e\t%U\t%S" -ao /data/chrpos.query.compressed.vcf.bcftools.time.txt \
    bcftools query -r "$query" -f '%LINE\n' /data/gwas.vcf.gz

    # compressed VCF using awk
    /usr/bin/time -f "%e\t%U\t%S" -ao /data/chrpos.query.compressed.vcf.awk.time.txt \
    gzip -dc /data/gwas.vcf.gz | \
    awk -v chr="$chr" -v pos="$pos" -F"\t" '$1 == chr && $2 == pos'

    # uncompressed VCF using awk
    /usr/bin/time -f "%e\t%U\t%S" -ao /data/chrpos.query.uncompressed.vcf.awk.time.txt \
    awk -v chr="$chr" -v pos="$pos" -F"\t" '$1 == chr && $2 == pos' /data/gwas.vcf

done < /data/queries.txt
```

Plot results

```{r}
library('ggplot2')
library('data.table')

# read in data
chrpos.query.uncompressed.text.awk.time <- fread("/data/chrpos.query.uncompressed.text.awk.time.txt")
chrpos.query.compressed.text.awk.time <- fread("/data/chrpos.query.compressed.text.awk.time.txt")
chrpos.query.compressed.vcf.bcftools.time <- fread("/data/chrpos.query.compressed.vcf.bcftools.time.txt")
chrpos.query.compressed.vcf.awk.time <- fread("/data/chrpos.query.compressed.vcf.awk.time.txt")
chrpos.query.uncompressed.vcf.awk.time <- fread("/data/chrpos.query.uncompressed.vcf.awk.time.txt")

chrpos.query.uncompressed.text.awk.time$method <- "uncompressed.text.awk"
chrpos.query.compressed.text.awk.time$method <- "compressed.text.awk"
chrpos.query.compressed.vcf.bcftools.time$method <- "compressed.vcf.bcftools"
chrpos.query.compressed.vcf.awk.time$method <- "compressed.vcf.awk"
chrpos.query.uncompressed.vcf.awk.time$method <- "uncompressed.vcf.awk"

# merge
all <- rbind(chrpos.query.uncompressed.text.awk.time[, c("real", "method")], chrpos.query.compressed.text.awk.time[, c("real", "method")], chrpos.query.compressed.vcf.bcftools.time[, c("real", "method")], chrpos.query.compressed.vcf.awk.time[, c("real", "method")], chrpos.query.uncompressed.vcf.awk.time[ ,c("real", "method")])
all$method <- as.factor(all$method)

# t test -- is there a query time difference uncompressed text and uncompressed VCF using awk?
t.test(chrpos.query.uncompressed.text.awk.time$real, chrpos.query.uncompressed.vcf.awk.time$real, paired = TRUE, alternative = "two.sided")

# plot
ggplot(all, aes(color = method, y = real)) + geom_boxplot() + labs(x = "Method", y = "Execution time (sec)") + theme(axis.text.x=element_blank()) + theme(legend.title=element_blank()) + ggtitle("Query time using chromosome position") + ylim(0, 6)
```

### Multitrait VCF file

File size and performance from storing multiple traits per VCF file

Make multisample VCF to approximate storing multiple traits in a single file

```{bash}
set -euo pipefail
bcftools merge --force-samples -O z -o /data/merged.vcf.gz /data/gwas.vcf.gz /data/gwas.vcf.gz /data/gwas.vcf.gz
bcftools index -t /data/merged.vcf.gz
```

One trait VCF file size

```{bash}
set -euo pipefail
ls -lh /data/gwas.vcf.gz
```

Three trait file size VCF

```{bash}
set -euo pipefail
ls -lh /data/merged.vcf.gz
```

Record time for random lookups using single and multi-trait VCF files

```{bash}
set -euo pipefail

echo -e "query\treal\tuser\tsys" > /data/chrpos.query.compressed.single.vcf.bcftools.time.txt
echo -e "query\treal\tuser\tsys" > /data/chrpos.query.compressed.multi.vcf.bcftools.time.txt

while read line; do

    # define query
    query=$(echo "$line" | cut -s -f2)

    # log query
    echo -ne "$query\t" >> /data/chrpos.query.compressed.single.vcf.bcftools.time.txt
    echo -ne "$query\t" >> /data/chrpos.query.compressed.multi.vcf.bcftools.time.txt

    # single trait
    /usr/bin/time -f "%e\t%U\t%S" -ao /data/chrpos.query.compressed.single.vcf.bcftools.time.txt \
    bcftools query -r "$query" -f '%LINE\n' /data/gwas.vcf.gz

    # multi trait
    /usr/bin/time -f "%e\t%U\t%S" -ao /data/chrpos.query.compressed.multi.vcf.bcftools.time.txt \
    bcftools query -r "$query" -f '%LINE\n' /data/merged.vcf.gz

done < /data/queries.txt
```

Plot results

```{r}
library('ggplot2')
library('data.table')

# read in data
chrpos.query.compressed.single.vcf.bcftools.time <- fread("/data/chrpos.query.compressed.single.vcf.bcftools.time.txt")
chrpos.query.compressed.multi.vcf.bcftools.time <- fread("/data/chrpos.query.compressed.multi.vcf.bcftools.time.txt")

chrpos.query.compressed.single.vcf.bcftools.time$method <- "compressed.single.vcf.bcftools"
chrpos.query.compressed.multi.vcf.bcftools.time$method <- "compressed.multi.vcf.bcftools"

# merge
all <- rbind(chrpos.query.compressed.single.vcf.bcftools.time[, c("real", "method")], chrpos.query.compressed.multi.vcf.bcftools.time[, c("real", "method")])
all$method <- as.factor(all$method)

# t test -- is there a query time difference between single & multisample VCF queries?
t.test(chrpos.query.compressed.single.vcf.bcftools.time$real, chrpos.query.compressed.multi.vcf.bcftools.time$real, paired = TRUE, alternative = "two.sided")

# plot
ggplot(all, aes(color = method, y = real)) + geom_boxplot() + labs(x = "Method", y = "Execution time (sec)") + theme(axis.text.x=element_blank()) + theme(legend.title=element_blank()) + ggtitle("Query time using coordinate for single vs multi-trait VCF") + ylim(0, 6)
```