---
title: "Map GWAS to VCF workflow"
author: "Matt Lyon"
date: "2020-02-25"
params:
    n_sim: 5
---

### Drop multialleleics and missing rsID field

Not including these variants in the performance comparison to avoid handling arrary values /querying on null values.

```{bash}
bcftools view \
-i 'ID != "." && COUNT(ALT) == 1 && COUNT(REF) == 1' \
-O z \
-o /data/gwas.vcf.gz \
/data/gwas.all.vcf.gz

bcftools index \
-t \
/data/gwas.vcf.gz
```

### Validate output

Check file format is valid. Genotype checking is skipped since the file has no genotypes.

```{bash}
set -euo pipefail
/usr/bin/gatk-4.1.5.0/gatk \
ValidateVariants \
-R /data/human_g1k_v37.fasta \
-V /data/gwas.vcf.gz \
--validation-type-to-exclude ALLELES
```

### Index on rsID

[rsidx](https://github.com/bioforensics/rsidx) is an excellent tool for indexing on VCF ID. This is made possible by extracting rsID and chromosome position to an SQLite database which is later queried to obtain the variant locus for extraction using tabix.

```{bash}
rsidx index \
/data/gwas.vcf.gz \
/data/gwas.vcf.gz.rsidx
```

### Create uncompressed VCF & index

Uncompress the file to see what the performance impact is later.

```{bash}
zcat /data/gwas.vcf.gz > /data/gwas.vcf
/usr/bin/gatk-4.1.5.0/gatk \
IndexFeatureFile \
-I /data/gwas.vcf
```

### Prepare tabular extract for performance comparisons

Create extract of VCF to text for performance comparisons later

```{bash}
echo -e "CHROM\tPOS\tREF\tALT\tID\tES\tSE\tLP\tAF\tSS" > /data/gwas.txt
bcftools query \
-f '%CHROM\t%POS\t%REF\t%ALT\t%ID[\t%ES\t%SE\t%LP\t%AF\t%SS]\n' \
/data/gwas.vcf.gz >> /data/gwas.txt
gzip -c /data/gwas.txt > /data/gwas.txt.gz
```

### Define queries

```{r}
library('data.table')
set.seed(12345)
options(scipen=999)

# read in tab GWAS
gwas <- fread("/data/gwas.txt")

# subsample variants
s <- gwas[sample(nrow(gwas), params$n_sim),]

# write out list of rsid for querying later
write.table(s$ID, sep="\t", row.names=F, quote=F, col.names=F, file="/data/rsid.txt")

# write out list of chrom pos for querying later
write.table(s[,c("CHROM", "POS")], sep="\t", row.names=F, quote=F, col.names=F, file="/data/chrpos.txt")

# write out list of 1Mb intervals and count number of variants for querying later
s$start <- s$POS - 500000
s$end <- s$POS + 500000
s$interval_count <- apply(s, 1, function(row){ length(which(gwas$CHROM == row[['CHROM']] & gwas$POS > as.numeric(row[['start']]) & gwas$POS < as.numeric(row[['end']]))) })
write.table(s[,c("CHROM", "start", "end", "interval_count")], sep="\t", row.names=F, quote=F, col.names=F, file="/data/intervals.txt")

# write out pval thresholds & counts for querying later
gwas$LP <- as.numeric(gwas$LP)

## define pval thresholds
thresholds <- seq(0.00000005, 1, 1/params$n_sim)
thresholds <- -log10(thresholds)

## round to avoid float imprecision issues
thresholds <- round(thresholds, 4)

## count records
counts <- sapply(thresholds, function (threshold) {sum(gwas$LP > threshold, na.rm = T)})
write.table(data.frame(thresholds=thresholds, counts=counts), sep="\t", row.names=F, quote=F, col.names=F, file="/data/pval.txt")

message(paste("Total number of variants in GWAS:", nrow(gwas)))
```
